test/integration/api/projects.js

```diff
@@ -1381,6 +1446,25 @@ describe('api: /projects?forms=true', () => {
             formList.length.should.equal(2);
           }))))));
 
+    it('should not return datasets to data collectors', testService(async (service, { Users }) => {
+      const asAlice = await service.login('alice');
+      const asChelsea = await service.login('chelsea');
+
+      await asAlice.post('/v1/projects/1/forms?publish=true')
+        .send(testData.forms.simpleEntity)
+        .set('Content-Type', 'application/xml')
+        .expect(200);
+
+      const chelsea = await Users.getByEmail('chelsea@getodk.org').then((o) => o.get());
+      await asAlice.post(`/v1/projects/1/assignments/formfill/${chelsea.actorId}`);
+
+      await asChelsea.get('/v1/projects?datasets=true')
+        .expect(200)
+        .then(({ body }) => {
+          body[0].datasetList.length.should.equal(0);
+        });
+    }));
+
```

Comment:
There is a `projects?forms=true` test below called `should set project data from formList even on non-extended projects` that should be duplicated for the dataset case. Does the `projects[i].datasets` and `projects[i].lastEntity` get updated appropriately? In particular, I think there's a bug in how the `lastEntity` value gets set at the whole project level (not within the dataset list within the project).